server/src/main/java/org/elasticsearch/rest/MethodHandlers.java

```diff
@@ -56,11 +58,16 @@ MethodHandlers addMethods(RestHandler handler, RestRequest.Method... methods) {
     }
 
     /**
-     * Returns the handler for the given method or {@code null} if none exists.
+     * Returns the handler for the given method and version or {@code null} if none exists.
      */
-    @Nullable
-    RestHandler getHandler(RestRequest.Method method) {
-        return methodHandlers.get(method);
+    RestHandler getHandler(RestRequest.Method method, Version version) {
+        Map<Version, RestHandler> versionToHandlers = methodHandlers.get(method);
+        if (versionToHandlers == null) {
+            return null; //method not found
+        }
+        final RestHandler handler = versionToHandlers.get(version);
+        return handler != null || version.equals(Version.CURRENT) ? handler : versionToHandlers.get(Version.CURRENT);
```

Comment:
I really can't remember the use cases where we needed this. (returning a v8 handler when a handler was not present in v7)
@jaymode   made a comment about this when we were still working on a feature branch https://github.com/elastic/elasticsearch/pull/54197/files#r414959451

Reply:
I think the reason for this is @Mpdreamz's comment https://github.com/elastic/elasticsearch/pull/60516#discussion_r500396954 where clients plan to send compatible with to all APIs. If this is the case, can we please document this in code/method so we do not lose the information

Reply:
I had to think about this waaay to long :) and wrote a unit test to help me understand
```
    /**
     * We want clients to be able to request compatibility with a prior version for all APIs. If a there does not exist a handler registered
     * for the prior version, we want to return the current version handler. This effectively means that the current endpoint is compatible
     * with the prior version, which should be true since if it were not compatible there would be a registered compatible endpoint.
     */
    public void testMissingPriorHandlerReturnsCurrentHandler(){
        RestHandler currentVersionHandler = new CurrentVersionHandler();
        MethodHandlers methodHandlers = new MethodHandlers("path", currentVersionHandler, RestRequest.Method.PUT, RestRequest.Method.POST);
        RestHandler handler = methodHandlers.getHandler(RestRequest.Method.PUT, Version.CURRENT.previousMajor());
        assertThat(handler, sameInstance(currentVersionHandler));
    }
```
Can you add a similar comment inline here, and add that test?

Reply:
The return can be simplified to:
```
        return handler == null ? versionToHandlers.get(Version.CURRENT) : handler;
```
with a test:
```
    public void testMissingCurrentHandler(){
        RestHandler previousVersionHandler = new PreviousVersionHandler();
        MethodHandlers methodHandlers = new MethodHandlers("path", previousVersionHandler, RestRequest.Method.PUT, RestRequest.Method.POST);
        RestHandler handler = methodHandlers.getHandler(RestRequest.Method.PUT, Version.CURRENT);
        assertNull(handler);
    }
```

Reply:
to be fair, I am not super convinced about returning the CURRENT.
if an endpoint did not exist in v7 server version, client would be getting 404 when trying to access it with v7 client.
When he upgrades server to next version and he still uses v7 client, I guess he should be still getting 404?


@jakelandis I am not sure I understand this:

> This effectively means that the current endpoint is compatible
>      * with the prior version, which should be true since if it were not compatible there would be a registered compatible endpoint.

from high level point of view, there is no knowledge if there is a compatible endpoint or not. If the handler's code did not change, user has no idea that a response came from v7 code, or if the response is `compatible shape` from v8.

THe same for when an endpoint do not exist. Returning 404 from v8 when for not existing in v7 endpoint is 100% compatible in shape and behaviour

@Mpdreamz any views on this?

Reply:
I think we should return new APIs with compatible requested. Below is a table with my thought for how behaviour should look. 

The rationale is that in a minor you can add a new API passively, therefor new APIs are compatible (as opposed to non-compatible/breaking). Further, one motivation to add a new API is to replace an old one. With a client sending compatibility all the time, they should be able to fix the compatibility warnings and migrate to the new API. This allows them to do just that. 

| scenario                                  | v7   | v8   | v7 compat / v8 server result |
| ---------------------------------------- | ---- | ---- | ---------------------------- |
| new in v8                                | 404  | 200  | 200 (v8)                     |
| no change in v8 (no compat handler)      | 200  | 200  | 200 (v8)                     |
| removed in v8 (with compat handler)      | 200  | 404  | 200 (v7)                     |
| non passive change (with compat handler) | 200  | 200  | 200 (v7)                     |

Reply:
> The rationale is that in a minor you can add a new API passively, therefor new APIs are compatible (as opposed to non-compatible/breaking)

great explanation. thank you. I think this was the initial reason we went for that behaviour on a feature branch. I will add a comment about this at least in a test

Reply:
Since clients generate methods for known API's the only way to talk to an API that exists in `8.0` but not (yet) in `7.x` is to call `transport.DoRequest(url)` 

I think the `new in v8` case needs some zooming in on:

# new in v8

| accept/content-type | v7 compat / v8 server result
| :--------------------| :--------------------------- |
| application/json    | application/json |
| application/vnd.elasticsearch+json;compatible-with=8 | application/vnd.elasticsearch+json;compatible-with=8 |
| application/vnd.elasticsearch+json;compatible-with=7 | ? 


The first two seem straight forward the last one not so much It could return a 

* 404, since thats what it did in 7
* 200 with `application/vnd.elasticsearch+json;compatible-with=7`. 
* 200 with `application/vnd.elasticsearch+json;compatible-with=8`. 

Returning with `compatible-with=8` is not what the user requested so seems like an unexpected return. 

Returning with `compatible-with=7` since its a new API it can be argued anything is compatible with `7`. But this will become a pain to track when e.g a new API gets introduced in 8.0 as experimental and GA's in `8.4`.  You potentially get back two different `7` responses for `8.0` and `8.4`. 

That leaves returning a 404.


cc @elastic/es-clients we have not discussed if the the transport used in the client should default to `application/vnd.elasticsearch+json;compatible-with=7` or `application/json`. I feel that if the transport defaults to `application/json` but the API methods calling the transport default to the vendor mime type the client user has the least suprises.

Reply:
I think with the current approach where we version compatible API `per major` we won't be able to give a 100% experience of previous version server. And that wasn't a goal - after all we want to achieve the compatible API, not the versioned one.
I feel all the comments from this "thread" are valid and I will summarise them on a separate issue. 
The current implementation is not returning the right response header yet, so the discussion on `new in v8` could continue on the new issue.
https://github.com/elastic/elasticsearch/issues/64852